physics/solar_system_body.hpp

```diff
@@ -81,26 +84,50 @@ void SolarSystem<Frame>::Initialize(
 
   // Store the data in maps keyed by body name.
   for (auto& body : *gravity_model_.mutable_gravity_model()->mutable_body()) {
-    auto const inserted =
+    bool inserted;
+    std::tie(std::ignore, inserted) =
         gravity_model_map_.insert(std::make_pair(body.name(), &body));
-    CHECK(inserted.second);
-  }
-  for (auto const& body : initial_state_.initial_state().body()) {
-    auto const inserted =
-        initial_state_map_.insert(std::make_pair(body.name(), &body));
-    CHECK(inserted.second);
+    CHECK(inserted) << body.name();
   }
+  if (initial_state_.initial_state().has_cartesian()) {
+    for (auto const& body : initial_state_.initial_state().cartesian().body()) {
+      bool inserted;
+      std::tie(std::ignore, inserted) =
+          cartesian_initial_state_map_.emplace(body.name(), &body);
+      CHECK(inserted) << body.name();
+    }
+
+    // Check that the maps are consistent.
+    auto it1 = gravity_model_map_.begin();
+    auto it2 = cartesian_initial_state_map_.begin();
+    for (; it1 != gravity_model_map_.end() &&
+           it2 != cartesian_initial_state_map_.end();
+         ++it1, ++it2) {
+      CHECK_EQ(it1->first, it2->first);
+      names_.push_back(it1->first);
+    }
+    CHECK(it1 == gravity_model_map_.end()) << it1->first;
+    CHECK(it2 == cartesian_initial_state_map_.end()) << it2->first;
```

Comment:
UB in my code?  This is actually safe, and the check as written is more useful than a size check as it prints out the offending body.